<h1>Chapter 4 | Data Exercise #1 | <code>hotels-vienna</code> | Comparison and correlation</h1>
<h2>Introduction:</h2>
<p>In this notebook, you will find my notes and code for Chapter 5's <b>exercise 1</b> of the book <a href="https://gabors-data-analysis.com/">Data Analysis for Business, Economics, and Policy</a>, by Gábor Békés and Gábor Kézdi. The question was: 
<p>1. Are central hotels better? To answer this, using the <code>hotels-vienna</code> dataset (as discussed in Chapter 3, Sesction 3.A1), create <b>two categories</b> by the <b>distance from center</b>: close and far (by picking a cutoff of your choice).</p>
<p>Assignments:</p>
<ul>
    <li>Show summary statistics.</li>
    <li>Compare <b>stars</b> and <b>ratings</b> and <b>prices</b> for close and far hotels.</li>
    <li>Create stacked bar charts, box plots, and violin plots.</li>
    <li>Summarize your findings.</li>
</ul>
<h2><b>1.</b> Load the data</h2>

In [19]:
import os
import sys
import warnings
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from plotnine import *
from mizani.formatters import percent_format

warnings.filterwarnings("ignore")
%matplotlib inline

In [20]:
# Increase number of returned rows in pandas
pd.set_option("display.max_rows", 500)

In [21]:
# Current script folder
current_path = os.getcwd()
dirname = current_path.split("da_data_exercises")[0]

# Get location folders
data_in = f"{dirname}da_data_repo/hotels-vienna/clean/"
data_out = f"{dirname}da_data_exercises/ch04-comparison_correlation/01-hotels_vienna/data/clean/"
output = f"{dirname}da_data_exercises/ch04-comparison_correlation/01-hotels_vienna/"
func = f"{dirname}da_case_studies/ch00-tech_prep/"
sys.path.append(func)

In [22]:
from py_helper_functions import *

In [23]:
df = pd.read_csv(f"{data_in}hotels-vienna.csv")

In [24]:
df.head()

country city_actual  rating_count center1label center2label neighbourhood  \
0  Austria      Vienna          36.0  City centre    Donauturm   17. Hernals   
1  Austria      Vienna         189.0  City centre    Donauturm   17. Hernals   
2  Austria      Vienna          53.0  City centre    Donauturm    Alsergrund   
3  Austria      Vienna          55.0  City centre    Donauturm    Alsergrund   
4  Austria      Vienna          33.0  City centre    Donauturm    Alsergrund   

   price    city  stars  ratingta  ...     offer_cat  year  month  weekend  \
0     81  Vienna    4.0       4.5  ...  15-50% offer  2017     11        0   
1     81  Vienna    4.0       3.5  ...   1-15% offer  2017     11        0   
2     85  Vienna    4.0       3.5  ...  15-50% offer  2017     11        0   
3     83  Vienna    3.0       4.0  ...  15-50% offer  2017     11        0   
4     82  Vienna    4.0       3.5  ...  15-50% offer  2017     11        0   

  holiday  distance  distance_alter  accommodation_type  nnights  rating  
0       0       2.7             4.4           Apartment        1     4.4  
1       0       1.7             3.8               Hotel        1     3.9  
2       0       1.4             2.5               Hotel        1     3.7  
3       0       1.7             2.5               Hotel        1     4.0  
4       0       1.2             2.8               Hotel        1     3.9  

[5 rows x 24 columns]

<p>Let's filter our data and restrict the accomodation type to hotels only.</p>

In [25]:
vienna_cut = df.loc[lambda x: x["accommodation_type"] == "Hotel"]

In [26]:
vienna_cut["accommodation_type"].value_counts()

Hotel    264
Name: accommodation_type, dtype: int64

<h2><b>2.</b> Binning quantitative <code>distance</code></h2>
<p>In chapter 3, we had defined that hotels above a 8-mile distance from the center were too far out and therefore marked as such. Given that a great deal of hotels were within a 2-mile distance, we can bin any hotel below this value as <b>close</b> and those farther than this distance as <b>far</b>. Let's do it.</p>

In [27]:
vienna_cut["distance2bins"] = np.where(vienna_cut["distance"] < 2, 1, pd.np.nan)
vienna_cut["distance2bins"] = np.where(vienna_cut["distance"] >= 2, 2, vienna_cut["distance2bins"])


In [28]:
vienna_cut.head()

country city_actual  rating_count center1label center2label neighbourhood  \
1  Austria      Vienna         189.0  City centre    Donauturm   17. Hernals   
2  Austria      Vienna          53.0  City centre    Donauturm    Alsergrund   
3  Austria      Vienna          55.0  City centre    Donauturm    Alsergrund   
4  Austria      Vienna          33.0  City centre    Donauturm    Alsergrund   
6  Austria      Vienna          57.0  City centre    Donauturm    Alsergrund   

   price    city  stars  ratingta  ...  year  month  weekend  holiday  \
1     81  Vienna    4.0       3.5  ...  2017     11        0        0   
2     85  Vienna    4.0       3.5  ...  2017     11        0        0   
3     83  Vienna    3.0       4.0  ...  2017     11        0        0   
4     82  Vienna    4.0       3.5  ...  2017     11        0        0   
6    103  Vienna    4.0       3.5  ...  2017     11        0        0   

  distance  distance_alter  accommodation_type  nnights  rating  distance2bins  
1      1.7             3.8               Hotel        1     3.9            1.0  
2      1.4             2.5               Hotel        1     3.7            1.0  
3      1.7             2.5               Hotel        1     4.0            1.0  
4      1.2             2.8               Hotel        1     3.9            1.0  
6      0.9             2.4               Hotel        1     3.9            1.0  

[5 rows x 25 columns]

In [15]:
vienna_cut["accommodation_type"].value_counts()

Hotel    264
Name: accommodation_type, dtype: int64

<h2><b>3.</b> Create summary statistics</b></h2>
<p>Let's create one table for each variable and compare the results.</p>

In [31]:
# Summary - stars
vienna_cut.filter(["stars","distance2bins"]).groupby("distance2bins").agg(
    ["min", "max", "mean","median", np.std, "size"]
)

stars                                     
                min  max      mean median       std size
distance2bins                                           
1.0             1.0  5.0  3.669192    4.0  0.777482  198
2.0             2.0  4.5  3.431818    3.0  0.600553   66

<p>In terms of stars, we can say that hotels that are close to the city center get more stars - 3.66 vs 3.43 for hotels that were qualified as far. But that is not the entire story. While hotels above the 2-mile threshold distance get a lower average score, there is <b>less uncertainty</b> regarding the spread of their data. Hotels closer to the center have lower min and higher max values, hence, you can expect to find a 1 to a 5-star hotel within this distance. Hotels farther from the center, meanwhile, have a higher min stars and a slightly lower max stars score - 2 and 4.5, respectively. As such, the standard deviation of hotels closer to center is higher, which reflects a higher degree of uncertainty.</p>
<p>Let's take a look at the ratings for each bin.</p>

In [32]:
# Summary - stars
vienna_cut.filter(["rating","distance2bins"]).groupby("distance2bins").agg(
    ["min", "max", "mean","median", np.std, "size"]
)

rating                                     
                 min  max      mean median       std size
distance2bins                                            
1.0              1.0  4.9  4.109645    4.1  0.441771  198
2.0              2.0  4.5  3.859091    4.0  0.471977   66

<p>Observations:</p>
<ul>
<li>The mean rating value for closer hotels is higher - 4.10 vs. 3.85.</li>
<li>The range of 

In [18]:
vienna_cut.head()

country city_actual  rating_count center1label center2label neighbourhood  \
1  Austria      Vienna         189.0  City centre    Donauturm   17. Hernals   
2  Austria      Vienna          53.0  City centre    Donauturm    Alsergrund   
3  Austria      Vienna          55.0  City centre    Donauturm    Alsergrund   
4  Austria      Vienna          33.0  City centre    Donauturm    Alsergrund   
6  Austria      Vienna          57.0  City centre    Donauturm    Alsergrund   

   price    city  stars  ratingta  ...  year  month  weekend  holiday  \
1     81  Vienna    4.0       3.5  ...  2017     11        0        0   
2     85  Vienna    4.0       3.5  ...  2017     11        0        0   
3     83  Vienna    3.0       4.0  ...  2017     11        0        0   
4     82  Vienna    4.0       3.5  ...  2017     11        0        0   
6    103  Vienna    4.0       3.5  ...  2017     11        0        0   

  distance  distance_alter  accommodation_type  nnights  rating  emp2bins  
1      1.7             3.8               Hotel        1     3.9       1.0  
2      1.4             2.5               Hotel        1     3.7       1.0  
3      1.7             2.5               Hotel        1     4.0       1.0  
4      1.2             2.8               Hotel        1     3.9       1.0  
6      0.9             2.4               Hotel        1     3.9       1.0  

[5 rows x 25 columns]